File: apps/api/src/Procraft.Application/Profiles/DTOs/ProfileDto.cs
using Procraft.Domain.Entities;

namespace Procraft.Application.Profiles.DTOs;

public sealed class ProfileDto
{
    public Guid Id { get; init; }
    public Guid UserId { get; init; }
    public Guid? TemplateId { get; init; }
    public string Username { get; init; } = string.Empty;
    public string FullName { get; init; } = string.Empty;
    public string? Title { get; init; }
    public string? Bio { get; init; }
    public string? Location { get; init; }
    public string? AvatarUrl { get; init; }
    public string TemplateSlug { get; init; } = "minimal";
    public DateTimeOffset CreatedAt { get; init; }
    public DateTimeOffset? UpdatedAt { get; init; }

    public IReadOnlyList<ProfileSkillDto> Skills { get; init; } = [];
    public IReadOnlyList<ProfileProjectDto> Projects { get; init; } = [];
    public IReadOnlyList<ProfileExperienceDto> WorkExperiences { get; init; } = [];
    public IReadOnlyList<ProfileEducationDto> Educations { get; init; } = [];
    public IReadOnlyList<ProfileCertificateDto> Certificates { get; init; } = [];
    public IReadOnlyList<ProfileSocialLinkDto> SocialLinks { get; init; } = [];

    public static ProfileDto FromProfile(Profile profile) =>
        new()
        {
            Id = profile.Id,
            UserId = profile.UserId,
            TemplateId = profile.TemplateId,
            Username = profile.User.Username,
            FullName = profile.FullName,
            Title = profile.Title,
            Bio = profile.Bio,
            Location = profile.Location,
            AvatarUrl = profile.AvatarUrl,
            TemplateSlug = profile.Template?.Slug ?? "minimal",
            CreatedAt = profile.CreatedAt,
            UpdatedAt = profile.UpdatedAt,
            Skills = profile.Skills.Select(s => new ProfileSkillDto(s.Id, s.Name, s.Level, s.Category)).ToList(),
            Projects = profile.Projects.Select(p => new ProfileProjectDto(p.Id, p.Name, p.Description, p.GithubUrl, p.IsRepositoryPrivate, p.LiveUrl)).ToList(),
            WorkExperiences = profile.WorkExperiences.Select(e => new ProfileExperienceDto(e.Id, e.Company, e.ExperienceType, e.Position, e.Description, e.StartDate, e.EndDate, e.IsCurrent)).ToList(),
            Educations = profile.Educations.Select(e => new ProfileEducationDto(e.Id, e.Institution, e.EducationType, e.Degree, e.Field, e.StartDate, e.EndDate)).ToList(),
            Certificates = profile.Certificates.Select(c => new ProfileCertificateDto(c.Id, c.Name, c.Issuer, c.IssuedOn, c.Url)).ToList(),
            SocialLinks = profile.SocialLinks.Select(s => new ProfileSocialLinkDto(s.Id, s.Platform, s.Url)).ToList(),
        };
}

public sealed record ProfileSkillDto(Guid Id, string Name, byte? Level, string? Category);
public sealed record ProfileProjectDto(Guid Id, string Name, string? Description, string? GithubUrl, bool IsRepositoryPrivate, string? LiveUrl);
public sealed record ProfileExperienceDto(Guid Id, string Company, string ExperienceType, string Position, string? Description, DateOnly? StartDate, DateOnly? EndDate, bool IsCurrent);
public sealed record ProfileEducationDto(Guid Id, string Institution, string EducationType, string? Degree, string? Field, DateOnly? StartDate, DateOnly? EndDate);
public sealed record ProfileCertificateDto(Guid Id, string Name, string? Issuer, DateOnly? IssuedOn, string? Url);
public sealed record ProfileSocialLinkDto(Guid Id, string Platform, string Url);

File: apps/api/src/profile/profile-dto.ts
import { toDateOnlyString, toDateTimeOffsetString } from '../common/dates';
import {
  CertificateEntity,
  EducationEntity,
  ProfileEntity,
  ProjectEntity,
  SkillEntity,
  SocialLinkEntity,
  TemplateEntity,
  UserEntity,
  WorkExperienceEntity,
} from '../database/entities';

export interface ProfileSections {
  skills: SkillEntity[];
  projects: ProjectEntity[];
  workExperiences: WorkExperienceEntity[];
  educations: EducationEntity[];
  certificates: CertificateEntity[];
  socialLinks: SocialLinkEntity[];
}

export const EMPTY_SECTIONS: ProfileSections = {
  skills: [],
  projects: [],
  workExperiences: [],
  educations: [],
  certificates: [],
  socialLinks: [],
};

/**
 * Port of Procraft.Application.Profiles.DTOs.ProfileDto.FromProfile. Mutation
 * endpoints intentionally return empty section arrays — matching the C#
 * handlers, which did not Include child collections on writes.
 */
export function toProfileDto(
  profile: ProfileEntity,
  user: UserEntity,
  template: TemplateEntity | null,
  sections: ProfileSections = EMPTY_SECTIONS,
) {
  return {
    id: profile.id,
    userId: profile.userId,
    templateId: profile.templateId ?? null,
    username: user.username,
    fullName: profile.fullName,
    title: profile.title ?? null,
    bio: profile.bio ?? null,
    location: profile.location ?? null,
    avatarUrl: profile.avatarUrl ?? null,
    templateSlug: template?.slug ?? 'minimal',
    createdAt: toDateTimeOffsetString(profile.createdAt),
    updatedAt: toDateTimeOffsetString(profile.updatedAt),
    skills: sections.skills.map((skill) => ({
      id: skill.id,
      name: skill.name,
      level: skill.level ?? null,
      category: skill.category ?? null,
    })),
    projects: sections.projects.map((project) => ({
      id: project.id,
      name: project.name,
      description: project.description ?? null,
      githubUrl: project.githubUrl ?? null,
      isRepositoryPrivate: project.isRepositoryPrivate,
      liveUrl: project.liveUrl ?? null,
    })),
    workExperiences: sections.workExperiences.map((experience) => ({
      id: experience.id,
      company: experience.company,
      experienceType: experience.experienceType,
      position: experience.position,
      description: experience.description ?? null,
      startDate: toDateOnlyString(experience.startDate),
      endDate: toDateOnlyString(experience.endDate),
      isCurrent: experience.isCurrent,
    })),
    educations: sections.educations.map((education) => ({
      id: education.id,
      institution: education.institution,
      educationType: education.educationType,
      degree: education.degree ?? null,
      field: education.field ?? null,
      startDate: toDateOnlyString(education.startDate),
      endDate: toDateOnlyString(education.endDate),
    })),
    certificates: sections.certificates.map((certificate) => ({
      id: certificate.id,
      name: certificate.name,
      issuer: certificate.issuer ?? null,
      issuedOn: toDateOnlyString(certificate.issuedOn),
      url: certificate.url ?? null,
    })),
    socialLinks: sections.socialLinks.map((link) => ({
      id: link.id,
      platform: link.platform,
      url: link.url,
    })),
  };
}
